lib/archive/diff.go

```diff
@@ -210,26 +240,54 @@ func createHeader(op trace.Operation, dir string, change docker.Change, spec *Fi
 		strippedName := strings.TrimPrefix(whiteOut, spec.StripPath)
 		hdr.Name = filepath.Join(spec.RebasePath, strippedName)
 	default:
-		fi, err := os.Lstat(filepath.Join(dir, change.Path))
+		path := filepath.Join(dir, change.Path)
+		fi, err := os.Lstat(path)
 		if err != nil {
 			op.Errorf("Error getting file info: %s", err.Error())
 			return nil, err
 		}
 
-		link, err := os.Readlink(filepath.Join(dir, change.Path))
-		if err != nil {
-			link = change.Path
+		link := ""
+		if fi.Mode()&os.ModeSymlink != 0 {
+			if link, err = os.Readlink(path); err != nil {
+				return nil, err
+			}
 		}
+
 		hdr, err = tar.FileInfoHeader(fi, link)
 		if err != nil {
 			op.Errorf("Error getting file info header: %s", err.Error())
 			return nil, err
 		}
 
-		change.Path = strings.TrimPrefix(change.Path, "/")
-		strippedName := strings.TrimPrefix(change.Path, spec.StripPath)
-		change.Path = strings.TrimPrefix(change.Path, "/")
-		hdr.Name = filepath.Join(spec.RebasePath, strippedName)
+		name := strings.TrimPrefix(change.Path, "/")
+		name = strings.TrimPrefix(name, spec.StripPath)
+		name = filepath.Join(spec.RebasePath, name)
+
+		if fi.IsDir() && !strings.HasSuffix(change.Path, "/") {
+			name += "/"
+		}
+
+		hdr.Name = strings.TrimPrefix(name, "/")
+
+		inode, err := setHeaderForSpecialDevice(hdr, fi.Sys())
```

Comment:
minor: it appears that we are using this function for a dual purpose here. Both to get the inode of the target as well as fixing the provided info. It might be more clear to provide the inode outside of this function. My thought on this is mainly because this function servers to correct the missing parts of the header for special devices and does not actually need the Inode at all to do this.